pkg/liquidity-source/dodo/classical/pool_simulator.go

```diff
@@ -0,0 +1,205 @@
+			Info: info,
+		},
+		Storage: poolState,
+		Tokens:  entity.ClonePoolTokens(entityPool.Tokens),
+		Meta:    meta,
+		gas:     DefaultGas,
+	}, nil
+}
+
+func (p *PoolSimulator) CalcAmountOut(param pool.CalcAmountOutParams) (*pool.CalcAmountOutResult, error) {
+	tokenAmountIn := param.TokenAmountIn
+	tokenOut := param.TokenOut
+
+	for i := range p.Info.Tokens {
+		if p.Info.Reserves[i].Cmp(big.NewInt(0)) <= 0 {
+			return &pool.CalcAmountOutResult{}, ErrReserveDepleted
+		}
+	}
+
+	if tokenAmountIn.Token != p.Info.Tokens[0] {
+		return &pool.CalcAmountOutResult{}, ErrOnlySupportSellBase
+	}
+
+	amountIn := number.SetFromBig(tokenAmountIn.Amount)
+
+	receiveQuote, lpFeeQuote, mtFeeQuote, _, _, _, err := p._querySellBaseToken(amountIn)
+	if err != nil {
+		return &pool.CalcAmountOutResult{}, err
+	}
+
+	fee := new(uint256.Int).Add(lpFeeQuote, mtFeeQuote)
+
+	return &pool.CalcAmountOutResult{
+		TokenAmountOut: &pool.TokenAmount{
+			Token:  tokenOut,
+			Amount: receiveQuote.ToBig(),
+		},
+		RemainingTokenAmountIn: &pool.TokenAmount{
+			Token:  tokenAmountIn.Token,
+			Amount: integer.Zero(),
+		},
+		Fee: &pool.TokenAmount{
+			Token:  tokenAmountIn.Token,
+			Amount: fee.ToBig(),
+		},
+		Gas: p.gas.SellBase,
+	}, nil
+}
+
+func (p *PoolSimulator) CalcAmountIn(param pool.CalcAmountInParams) (*pool.CalcAmountInResult, error) {
+	tokenIn := param.TokenIn
+	tokenAmountOut := param.TokenAmountOut
+
+	for i := range p.Info.Tokens {
+		if p.Info.Reserves[i].Cmp(big.NewInt(0)) <= 0 {
+			return &pool.CalcAmountInResult{}, ErrReserveDepleted
+		}
+	}
+
+	if tokenAmountOut.Token != p.Info.Tokens[0] {
+		return &pool.CalcAmountInResult{}, ErrOnlySupportBuyBase
+	}
+
+	amountOut := number.SetFromBig(tokenAmountOut.Amount)
+
+	payQuote, lpFeeBase, mtFeeBase, _, _, _, err := p._queryBuyBaseToken(amountOut)
+	if err != nil {
+		return &pool.CalcAmountInResult{}, err
+	}
+
+	fee := new(uint256.Int).Add(lpFeeBase, mtFeeBase)
+
+	return &pool.CalcAmountInResult{
+		TokenAmountIn: &pool.TokenAmount{
+			Token:  tokenIn,
+			Amount: payQuote.ToBig(),
+		},
+		RemainingTokenAmountOut: &pool.TokenAmount{
+			Token:  tokenAmountOut.Token,
+			Amount: integer.Zero(),
+		},
+		Fee: &pool.TokenAmount{
+			Token:  tokenIn,
+			Amount: fee.ToBig(),
+		},
+		Gas: p.gas.BuyBase,
+	}, nil
+}
+
+func (p *PoolSimulator) UpdateBalance(params pool.UpdateBalanceParams) {
+	input, output := params.TokenAmountIn, params.TokenAmountOut
+	var isSellBase bool
+	if input.Token == p.Info.Tokens[0] {
+		isSellBase = true
+	} else {
+		isSellBase = false
+	}
+	inputAmount := input.Amount
+	// output.Amount was already fee-deducted in CalcAmountOut above, need to add back to update balances
+	outputAmount := new(big.Int).Add(output.Amount, params.Fee.Amount)
+
+	if isSellBase {
+		// p.Info.Reserves[0] = p.Info.Reserves[0] + inputAmount
+		// p.Info.Reserves[1] = p.Info.Reserves[1] - outputAmount - mtFee
+		p.Info.Reserves[0] = new(big.Int).Add(p.Info.Reserves[0], inputAmount)
```

Comment:
We can save one allocation here.
```suggestion
		p.Info.Reserves[0].Add(p.Info.Reserves[0], inputAmount)
```